lib/libzpool/Makefile.am

```diff
@@ -22,6 +22,10 @@ USER_C = \
 	taskq.c \
 	util.c
 
+if BUILD_LINUX
+USER_C += os/linux/compat.c
```

Comment:
i'd like to propose do not move files to linux/other specific directory, try to use file like:
`compat_linux.c`
in the same place and include it by target `BUILD_LINUX ` where we need it
i'll try to use this logic with `compat_dilos.c` in this case
try to move files to `os/[linux|dilos|freebsd|others]/ `is not good to me right now with many manual merges and have all files in the one place is much more better

Reply:
That’s what I did initially. It makes it harder to segregate platforms. How does it benefit you?

Reply:
example:
```
ztypes_darwin_arm64.go
ztypes_dragonfly_amd64.go
ztypes_freebsd_386.go
ztypes_freebsd_amd64.go
ztypes_freebsd_arm.go
ztypes_linux_386.go
ztypes_linux_amd64.go
ztypes_linux_arm.go
ztypes_linux_arm64.go
ztypes_linux_mips.go
ztypes_linux_mips64.go
ztypes_linux_mips64le.go
ztypes_linux_mipsle.go
ztypes_linux_ppc64.go
ztypes_linux_ppc64le.go
ztypes_linux_s390x.go
ztypes_netbsd_386.go
ztypes_netbsd_amd64.go
ztypes_netbsd_arm.go
ztypes_netbsd_arm64.go
ztypes_openbsd_386.go
ztypes_openbsd_amd64.go
ztypes_openbsd_arm.go
ztypes_openbsd_arm64.go
ztypes_solaris_amd64.go```
all files in one directory - easy to manage and manual merge between platforms.
we can include platform specific file in Makefile based on platform identification.

Reply:
I _personally_ really don't like it. I'm approaching this from treating it the way different arches on support on FreeBSD and Linux.

```
/devel/linux$ ls arch/
alpha  arm    c6x   h8300    ia64     m68k        mips   nios2     parisc   riscv  sh     um         x86
arc    arm64  csky  hexagon  Kconfig  microblaze  nds32  openrisc  powerpc  s390   sparc  unicore32  xtensa
```
```
ls devel/freebsd/sys/
amd64  cam     contrib  dts   gnu   kgssapi   modules   netinet   netsmb     nlm         riscv     sys    ufs  xen
arm    cddl    crypto   fs    i386  libkern   net       netinet6  nfs        ofed        rpc       teken  vm
arm64  compat  ddb      gdb   isa   Makefile  net80211  netipsec  nfsclient  opencrypto  security  tests  x86
bsm    conf    dev      geom  kern  mips      netgraph  netpfil   nfsserver  powerpc     sparc64   tools  xdr
```

Reply:
i do a lot of manual merges by ARAXIS merge:
https://share.getcloudapp.com/z8uwrmA2
left panel is ZoL, right is dilos.
command with full path like:
`merge btree.c F:\rws1\users\igor\dilos-illumos.zfs\usr\src\uts\common\fs\zfs\btree.c`
https://share.getcloudapp.com/12uEvL9r

and it's very hard right now to do it, because i have to move to another directory with sources - os/linux/ - and try to look/merge another files what was moved from original location.
before relocation, it was much more better - one sources in one directory on left and merge with right place.
it is like usability for integration process where you do a lot of manual merges.
additional benefit - you can see all files and can split it to another platform and can checl every time - you no need change your place for the platform specific files.

Reply:
@ikozhukhov FreeBSD's legacy ZFS layout has a complete different structure in turn from ZoL (mirroring the layout in Illumos) but @behlendorf  has been able to track changes made there against changes made to ZoL. Maybe he has some pointers for you.

Reply:
no hints for manual merge where you have to provide full location to file.
`git cherry-pick <changeset>` can try to find file, but not every time.
i have illumos based files location now, but i have plans move to ZoL sources.
with new structure it will be not easy to check files by manual diff if i wan to do it.
i tried to learn current new structure, but it is not comfortable to me right now where i need more time merge changes from ZoL to dilos.
i'd like open topics for restructure of source tree while we are not so far on it.

Reply:
@ikozhukhov thanks for the feedback.  I see what you're getting at and how this could add some additional steps when manually merging patches.  However, I think this should be a relatively short term issue while we work through restructuring the repository.

The intention is that once everything has been split out the vast majority of the changes will only touch the common code.  Moreover, any platform specific changes will be easily identified since they'll appear in the relevant platform directories.

> i do a lot of manual merges by ARAXIS merge

Personally, I've had quite a lot of success using `git` to cherry-pick patches directly from FreeBSD and Illumos in to the ZoL source tree.  Hopefully, you'll be able to do something like this for dilos where you can add the ZoL/OpenZFS repository as a remote and automatically merge or cherry-pick the new changes.

Let's continue with the current refactoring and in parallel explore ways to improve your workflow.

Reply:
i'm not sure it will be short term in future, where some changes can be implemented in one platform and ported to another with delay. mainstream platform will be one and others can be affected and need a time for better port of changes.
will this case in many time we need to work with others platforms sources and understand how to port changes to another platform - this process will be time to time present to me, because no showstoppers for others platform to wait of big features integration and wants to go forward. i know this situation - it was the same with xen ports. but diff and merge sources in different directories - not good way. yes, i can try to resolve this issue on my place by symlink of files from another location to one directory with different platform specific name, but will be much more better to have agreements with one tree. and yes, i'm only one with this question/proposal with own platform and can try to do what i need, but it's not easy with git cherry-pick with conflicts where i need to do manual diff.

Reply:
I'm a bit more optimistic about the situation.  The plan of record is that any new feature being merged in to the shared repository needs to pass the CI for all of the support platforms.  Initially, that will be Linux and FreeBSD but we'll want to extend the CI to cover each platform as they're included in the shared repository.

Reply:
you are optimistic :) i'd like to be optimistic too, but based on practice of support of different platforms, it's very hard to be in sync a many time in platform specific use-cases. Example: VDEV identification. and many time i need to look implementation of platform specific updates in platform specific files and try to think how to adopt it to our own platform. it is why relocation of files can impact this situation with manual diff/merge. we try and will try prepare wrappers to shared functions, but need take a look implementations based on others platforms. platform specific implementations will be moved to platform specific directories and have to change current directory to look implementation/diff and go back for next merges. no every `git cherry-pick` can be done as is - many time i have to fix merge conflicts. yes, we try to minimize it by merge of features and implement differences, but it will be not done a very long time and right now it really overkill of my merge time with new files relocation, where i need to spend additional time for merges of git conflicts.